File: Project/Python/project/components/store/dave_store.py
# -*- coding: utf-8 -*-
#/*
# * Copyright (c) 2023 Renwei
# *
# * This is a free software; you can redistribute it and/or modify
# * it under the terms of the MIT license. See LICENSE for details.
# */
from public import *
from public.base.dave_log import DAVELOG


STORE_THREAD_NAME=b"store"


# =====================================================================


def STORESQL(*sql: object):
    sql = str(sql[0])

    pReq = thread_msg(StoreMysqlReq)
    pReq.contents.sql = str_to_mbuf(sql)
    pReq.contents.ptr = None

    pRsp = write_co(STORE_THREAD_NAME, STORE_MYSQL_REQ, pReq, STORE_MYSQL_RSP, StoreMysqlRsp)
    if pRsp == None:
        DAVELOG(f"pRsp is None / sql:{sql}")
        return ERRCODE_ptr_null, None

    if (pRsp.ret != RetCode_OK) and (pRsp.ret != RetCode_empty_data) and (pRsp.ret != RetCode_table_exist):
        DAVELOG(f"ret:{t_auto_RetCode_str(pRsp.ret)}/{pRsp.msg}, sql:{sql}")
        return pRsp.ret, None

    sql_array = mbuf_to_dict(pRsp.data)

    dave_mfree(pRsp.data)

    return RetCode_OK, sql_array


def STORELOAD(sql_array, row, column):
    if sql_array == None:
        return None
    if row >= len(sql_array):
        DAVELOG(f"sql_array:{sql_array} row:{row} is None")
        return None
    if column >= len(sql_array[row]):
        DAVELOG(f"sql_array:{sql_array} row:{row} column:{column} is None")
        return None

    return sql_array[row][column]


def STORELOADStr(sql_array, column):
    if sql_array == None:
        return None

    str_str = STORELOAD(sql_array, 0, column)
    if str_str == None:
        DAVELOG(f"sql_array:{sql_array} column:{column} is None")
        return None
    return str_str


def STORESQLCREATETABLE(db_name, table_name, table_disc):
    _, data_array = STORESQL(f"SHOW TABLES IN {db_name} LIKE '{table_name}'")
    if (data_array != None) and len(data_array) != 0:
        return
    STORESQL(f"CREATE TABLE {db_name}.{table_name} {table_disc}")
    return


def STORESQLCLEANTABLE(db_name, table_name, table_disc):
    _, data_array = STORESQL(f"SHOW TABLES IN {db_name} LIKE '{table_name}'")
    if (data_array != None) and len(data_array) != 0:
        STORESQL(f"DROP TABLE {db_name}.{table_name}")
        DAVELOG(f"STORESQLCLEANTABLE: DROP TABLE {db_name}.{table_name}")
    STORESQL(f"CREATE TABLE {db_name}.{table_name} {table_disc}")
    return

File: Project/Go/project/components/store/dave_store.go
package store

/*
 * Copyright (c) 2023 Renwei
 *
 * This is a free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

import (
	"dave/public/base"
	"dave/public/auto"
	"dave/public/tools"
	"fmt"
	"unsafe"
	"errors"
	"strconv"
)

const STORE_THREAD_NAME = "store"

func _store_sql_co(sql_string string) (*auto.StoreMysqlRsp, error){
	req := auto.StoreMysqlReq{}
	req.Sql = base.T_gostring2mbuf(sql_string)
	req.Ptr = 0

	pRsp := (*auto.StoreMysqlRsp)(base.Write_co(STORE_THREAD_NAME, auto.STORE_MYSQL_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req), auto.STORE_MYSQL_RSP))
	if pRsp == nil {
		return nil, errors.New("co timer out")
	}

	return pRsp, nil
}

func _store_sql_sync(sql_string string) (*auto.StoreMysqlRsp, error){
	req := auto.StoreMysqlReq{}
	req.Sql = base.T_gostring2mbuf(sql_string)
	req.Ptr = 0

	rsp := auto.StoreMysqlRsp{}

	ret := base.Sync_msg(STORE_THREAD_NAME, auto.STORE_MYSQL_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req), auto.STORE_MYSQL_RSP, int(unsafe.Sizeof(rsp)), unsafe.Pointer(&rsp))
	if ret == false {
		return nil, errors.New("sync error!")
	}

	return &rsp, nil
}

// =====================================================================

func STORESQL(format string, sql ...interface{}) (*tools.Json, error) {
	sql_string := fmt.Sprintf(format, sql...)

	pRsp, err := _store_sql_sync(sql_string)
	if err != nil {
		return nil, err
	}

	if (pRsp.Ret != auto.RetCode_OK) && 
		(pRsp.Ret != auto.RetCode_empty_data) &&
		(pRsp.Ret != auto.RetCode_table_exist) {

		base.DAVELOG("ret:%s on sql:%s", auto.T_auto_RetCode_str(pRsp.Ret), sql_string)

		base.Dave_mfree(pRsp.Data)

		return nil, errors.New("sql failed")
	}

	json_obj, _ := base.T_mbuf2json(pRsp.Data)

	base.Dave_mfree(pRsp.Data)

	base.DAVEDEBUG("sql:%v ret:%v", sql_string, json_obj)

	return json_obj, nil
}

func STORELOAD(json_obj *tools.Json, row int, column int) *tools.Json {
	return json_obj.GetIndex(row).GetIndex(column)
}

func STORELOADRStr(json_obj *tools.Json, row int, column int) string {
	if json_obj == nil {
		return ""
	}

	str_str, err := STORELOAD(json_obj, row, column).String()
	if err != nil {
		base.DAVEDEBUG("err:%v column:%v json:%v", err, column, json_obj)
		return ""
	}
	return str_str
}

func STORELOADStr(json_obj *tools.Json, column int) string {
	if json_obj == nil {
		return ""
	}

	str_str, err := STORELOAD(json_obj, 0, column).String()
	if err != nil {
		base.DAVEDEBUG("err:%v column:%v json:%v", err, column, json_obj)
		return ""
	}
	return str_str
}

func STORELOADSb(json_obj *tools.Json, column int) int {
	if json_obj == nil {
		return 0
	}

	int_str, err := STORELOAD(json_obj, 0, column).String()
	if err != nil {
		base.DAVEDEBUG("err:%v column:%v json:%v", err, column, json_obj)
		return 0
	}
	int_value, _ := strconv.ParseInt(int_str, 10, 64)
	return int(int_value)
}